### PR TITLE
Device profile dynamic value

### PR DESCRIPTION
**Pull request by AndrewVolosytnykhThingsboard**

### CODE REVIEW

**Comment by AndrewVolosytnykhThingsboard** [comment]
@vvlladd28 Can you please review code and specify how can I improve solution for front-end ?

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 1]
```diff
@@ -480,7 +488,41 @@ public void upgradeDatabase(String fromVersion) throws Exception {
                     } catch (Exception e) {
                         log.error("Failed updating schema!!!", e);
                     }
+                  
+                    deviceProfileRepository.findAll().forEach(deviceProfile -> {
+                        if(deviceProfile.getProfileData().has("alarms")) {
+                            JsonNode array = deviceProfile.getProfileData().get("alarms");
+                            for (JsonNode node : array) {
+                                if (node.has("createRules")) {
+                                    JsonNode createRules = node.get("createRules");
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:497]: You can iterate through all the keys. OR at least use Severity enum. This is terrible code duplication.

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:497]: What if you missed one of the severities?

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 2]
```diff
@@ -480,7 +488,41 @@ public void upgradeDatabase(String fromVersion) throws Exception {
                     } catch (Exception e) {
                         log.error("Failed updating schema!!!", e);
                     }
+                  
+                    deviceProfileRepository.findAll().forEach(deviceProfile -> {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:492]: findAll is very aggressive. We have environments with more than 50K device profiles. Please use pagination. 
Move this code to DefaultDataUpdateService (see tenantsDefaultRuleChainUpdater or similar method there)

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 3]
```diff
@@ -449,38 +453,35 @@ public void upgradeDatabase(String fromVersion) throws Exception {
             case "3.2.2":
                 try (Connection conn = DriverManager.getConnection(dbUrl, dbUserName, dbPassword)) {
                     log.info("Updating schema ...");
+                    conn.createStatement().execute("CREATE TABLE IF NOT EXISTS resource ( " +
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:456]: Why this lines are affected by your PR?

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 4]
```diff
@@ -97,6 +98,9 @@
     @Autowired
     private DeviceProfileService deviceProfileService;
 
+    @Autowired
+    private DeviceProfileRepository deviceProfileRepository;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:102]: You no longer need this change here.

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java (hunk 3 above)
Comment (AndrewVolosytnykhThingsboard) [reply application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:456]: @ashvayka  I had deleted redundant "try-catch"

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:451]: They are not redundant. Don't touch the code that is not related to your changes, please. This block merge of main changes.

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 5]
```diff
@@ -261,4 +303,28 @@ private void updateTenantEntityViews(TenantId tenantId) {
         }, MoreExecutors.directExecutor());
     }
 
+    private void convertOldSpecToNew(JsonNode spec) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:306]: please rename to convertDeviceProfileAlarmRulesForVersion330

File: application/src/main/java/org/thingsboard/server/service/install/update/EntityPaginatedUpdater.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.thingsboard.server.common.data.SearchTextBased;
+import org.thingsboard.server.common.data.id.UUIDBased;
+import org.thingsboard.server.common.data.page.PageLink;
+import org.thingsboard.server.dao.DaoUtil;
+import org.thingsboard.server.dao.model.BaseSqlEntity;
+
+public abstract class EntityPaginatedUpdater<I, D extends BaseSqlEntity<? extends SearchTextBased<? extends UUIDBased>>> {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/EntityPaginatedUpdater.java:26]: Why not just change from
```java
public abstract class PaginatedUpdater<I, D extends SearchTextBased<? extends UUIDBased>> {
```
to 
```java 
public abstract class PaginatedUpdater<I, D> {
```

and execute DaoUtil.toPageable(pageLink) inside the findEntities..

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 7]
```diff
@@ -114,13 +124,50 @@ public void updateData(String fromVersion) throws Exception {
...
+                                    if (createRules.has(severity.name())) {
+                                        convertDeviceProfileAlarmRulesForVersion330(createRules.get(severity.name()).get("condition").get("spec"));
+                                    }
+                                }
+                            }
+                            if (node.has("clearRule") && !node.get("clearRule").isNull()) {
+                                convertDeviceProfileAlarmRulesForVersion330(node.get("clearRule").get("condition").get("spec"));
+                            }
+                        }
+                        deviceProfileRepository.save(deviceProfile);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:166]: don't save the object if there are no changes.

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 8]
```diff
@@ -370,6 +417,30 @@ private void updateTenantAlarmsCustomer(TenantId tenantId) {
         }
     }
 
+    private void convertDeviceProfileAlarmRulesForVersion330(JsonNode spec) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:420]: make this function to return boolean (changed or not)

**Comment by ashvayka** [comment]
please fix two minor comments for the upgrade procedure and we will merge.

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 9]
```diff
@@ -100,7 +100,6 @@
     @Autowired
     private ApiUsageStateService apiUsageStateService;
 
-
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:103]: remove unnecessary changes

File: application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java [hunk 10]
```diff
@@ -495,4 +494,5 @@ protected boolean isOldSchema(Connection conn, long fromVersion) {
         }
         return isOldSchema;
     }
+
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java:497]: remove unnecessary changes
